File: solution/leetcode/239.cpp
// https://leetcode.com/problems/sliding-window-maximum/
// Time: O(n) each element is pushed/poped in deque only once, Space: O(k)

class Solution {
	void insert(deque<pair<int, int>> &q, int val, int index) {
		while (q.empty() == false && q.back().first < val)
			q.pop_back();
		q.push_back({val, index});
	}

	void remove(deque<pair<int, int>> &q, int val, int index) {
		if (q.empty() || q.front().second > index)
			return;
		else if (q.front().second == index)
			q.pop_front();
	}

public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    	deque<pair<int, int>> q;
    	vector<int> ans;
    	if (nums.size() < k)
    		return {};

    	for (int i = 0; i < k; i += 1) 
    		insert(q, nums[i], i);
    	ans.push_back(q.front().first);

    	for (int i = k; i < nums.size(); i += 1) {
    		remove(q, nums[i-k], i-k);
    		insert(q, nums[i], i);
    		ans.push_back(q.front().first);
    	}

    	return ans;
    }
};

File: solution/leetcode/239.py
# https://leetcode.com/problems/sliding-window-maximum/description/


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        stack, ans = [], []
        # initialize the monotonic stack (index, val)
        # the largest is present at the earliest
        for index, num in enumerate(nums[:k]):
            self.addToStack(stack, index, num)
        
        ans.append(stack[0][1])
        for index, num in enumerate(nums[k:], k):
            if stack[0][0] <= index-k:
                stack = stack[1:]
            self.addToStack(stack, index, num)
            ans.append(stack[0][1])
        
        return ans

    def addToStack(self, stack, index, num):
        while stack and stack[-1][1] <= num:
            stack.pop()
        if not stack or stack[-1][1] > num:
            stack.append((index, num))
        
